File: examples/ocaml-functional/option_map.ml
type 'a option_value =
  | None
  | Some of 'a

let increment value = value + 1

let map_option function_value option =
  match option with
  | None -> None
  | Some value -> Some (function_value value)

let gpu_main =
  match map_option increment (Some 41) with
  | None -> 0
  | Some value -> value

File: examples/rust-functional/option_map.rs
enum Option<T> {
    None,
    Some(T),
}

fn increment(value: i32) -> i32 {
    value + 1
}

fn map<A, B>(function: fn(A) -> B, value: Option<A>) -> Option<B> {
    match value {
        Option::None => Option::None,
        Option::Some(inner) => Option::Some(function(inner)),
    }
}

fn gpu_main() -> i32 {
    match map(increment, Option::Some(41)) {
        Option::None => 0,
        Option::Some(value) => value,
    }
}

fn main() {}
